File: app/src/main/java/cl/monsoon/s1next/view/fragment/GeneralPreferenceFragment.java
package cl.monsoon.s1next.view.fragment;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.v7.preference.Preference;

import javax.inject.Inject;

import cl.monsoon.s1next.App;
import cl.monsoon.s1next.R;
import cl.monsoon.s1next.data.event.FontSizeChangeEvent;
import cl.monsoon.s1next.data.event.ThemeChangeEvent;
import cl.monsoon.s1next.data.pref.GeneralPreferencesManager;
import cl.monsoon.s1next.data.pref.ThemeManager;
import cl.monsoon.s1next.util.DeviceUtil;
import cl.monsoon.s1next.util.ResourceUtil;
import cl.monsoon.s1next.view.activity.SettingsActivity;
import cl.monsoon.s1next.widget.EventBus;

/**
 * An Activity includes general settings that allow users
 * to modify general features and behaviors such as theme
 * and font size.
 */
public final class GeneralPreferenceFragment extends BasePreferenceFragment
        implements Preference.OnPreferenceClickListener {

    public static final String PREF_KEY_THEME = "pref_key_theme";
    public static final String PREF_KEY_FONT_SIZE = "pref_key_font_size_v2";
    public static final String PREF_KEY_SIGNATURE = "pref_key_signature";

    private static final String PREF_KEY_DOWNLOADS = "pref_key_downloads";

    @Inject
    EventBus mEventBus;

    @Inject
    GeneralPreferencesManager mGeneralPreferencesManager;

    @Inject
    ThemeManager mThemeManager;

    @Override
    public void onCreatePreferences(Bundle bundle, String s) {
        addPreferencesFromResource(R.xml.preference_general);
        App.getAppComponent(getContext()).inject(this);

        findPreference(PREF_KEY_DOWNLOADS).setOnPreferenceClickListener(this);
        findPreference(PREF_KEY_SIGNATURE).setSummary(DeviceUtil.getSignature(getContext()));
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        switch (key) {
            case PREF_KEY_THEME:
                mThemeManager.invalidateTheme();
                mEventBus.post(new ThemeChangeEvent());

                break;
            case PREF_KEY_FONT_SIZE:
                mGeneralPreferencesManager.invalidateFontScale();
                // change scaling factor for fonts
                ResourceUtil.setScaledDensity(getResources(),
                        mGeneralPreferencesManager.getFontScale());
                mEventBus.post(new FontSizeChangeEvent());

                break;
            case PREF_KEY_SIGNATURE:
                mGeneralPreferencesManager.invalidateSignatureEnabled();

                break;
            default:
                // fall through
        }
    }

    @Override
    public boolean onPreferenceClick(Preference preference) {
        switch (preference.getKey()) {
            case PREF_KEY_DOWNLOADS:
                SettingsActivity.startDownloadSettingsActivity(preference.getContext());

                return true;
            default:
                // fall through
        }

        return false;
    }
}

File: app/src/main/java/me/ykrank/s1next/view/page/setting/fragment/GeneralPreferenceFragment.kt
package me.ykrank.s1next.view.page.setting.fragment

import android.content.SharedPreferences
import android.os.Bundle
import androidx.lifecycle.lifecycleScope
import androidx.preference.Preference
import com.github.ykrank.androidautodispose.AndroidRxDispose
import com.github.ykrank.androidlifecycle.event.FragmentEvent
import com.github.ykrank.androidtools.util.L
import com.github.ykrank.androidtools.util.ResourceUtil
import com.github.ykrank.androidtools.util.RxJavaUtil
import com.github.ykrank.androidtools.widget.EventBus
import com.github.ykrank.androidtools.widget.track.DataTrackAgent
import com.github.ykrank.androidtools.widget.track.event.ThemeChangeTrackEvent
import io.reactivex.Single
import io.reactivex.functions.Consumer
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import me.ykrank.s1next.App
import me.ykrank.s1next.R
import me.ykrank.s1next.data.pref.GeneralPreferencesManager
import me.ykrank.s1next.data.pref.ThemeManager
import me.ykrank.s1next.util.AppDeviceUtil
import me.ykrank.s1next.util.AppUpdate
import me.ykrank.s1next.util.BuglyUtils
import me.ykrank.s1next.view.event.FontSizeChangeEvent
import me.ykrank.s1next.view.event.ThemeChangeEvent
import me.ykrank.s1next.view.page.setting.SettingsActivity
import me.ykrank.s1next.widget.span.HtmlCompat
import me.ykrank.s1next.widget.span.HtmlCompat.FROM_HTML_MODE_LEGACY
import javax.inject.Inject

/**
 * An Activity includes general settings that allow users
 * to modify general features and behaviors such as theme
 * and font size.
 */
class GeneralPreferenceFragment : BasePreferenceFragment(), Preference.OnPreferenceClickListener {

    @Inject
    internal lateinit var mEventBus: EventBus

    @Inject
    internal lateinit var mGeneralPreferencesManager: GeneralPreferencesManager

    @Inject
    internal lateinit var mThemeManager: ThemeManager

    @Inject
    internal lateinit var trackAgent: DataTrackAgent

    override fun onCreatePreferences(bundle: Bundle?, s: String?) {
        App.appComponent.inject(this)
        addPreferencesFromResource(R.xml.preference_general)

        findPreference<Preference>(getString(R.string.pref_key_downloads))?.onPreferenceClickListener = this
        findPreference<Preference>(getString(R.string.pref_key_blacklists))?.onPreferenceClickListener = this
        findPreference<Preference>(getString(R.string.pref_key_post_read))?.onPreferenceClickListener = this
        findPreference<Preference>(getString(R.string.pref_key_backup))?.onPreferenceClickListener = this
        findPreference<Preference>(getString(R.string.pref_key_network))?.onPreferenceClickListener = this

        if (BuglyUtils.isPlay) {
            findPreference<Preference>(getString(R.string.pref_key_check_update))?.isVisible = false
        } else {
            findPreference<Preference>(getString(R.string.pref_key_check_update))?.onPreferenceClickListener = this
        }

        updateSignature()
    }

    fun updateSignature() {
        lifecycleScope.launch {
            val signature = if (mGeneralPreferencesManager.isDeviceInfoShownInSignature) {
                AppDeviceUtil.getSignatureWithDeviceInfo(requireContext())
            } else {
                AppDeviceUtil.getSignature(requireContext())
            }
            val html = withContext(Dispatchers.Default) {
                HtmlCompat.fromHtml(signature, FROM_HTML_MODE_LEGACY)
            }
            findPreference<Preference>(getString(R.string.pref_key_signature))?.summary = html
        }
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences?, key: String?) {
        if (!isAdded) {
            return
        }
        if (key == getString(R.string.pref_key_theme) || key == getString(R.string.pref_key_dark_theme)) {
            trackAgent.post(ThemeChangeTrackEvent(false))
            mThemeManager.invalidateTheme()
            mEventBus.postDefault(ThemeChangeEvent())
        } else if (key == getString(R.string.pref_key_font_size)) {
            L.l("Setting")
            // change scaling factor for fonts
            ResourceUtil.setScaledDensity(
                activity,
                mGeneralPreferencesManager.fontScale
            )
            mEventBus.postDefault(FontSizeChangeEvent())
        } else if (key == getString(R.string.pref_key_device_info)) {
            updateSignature()
        }
    }

    override fun onPreferenceClick(preference: Preference): Boolean {
        val key = preference.key ?: return false
        when (key) {
            getString(R.string.pref_key_downloads) -> {
                SettingsActivity.startDownloadSettingsActivity(preference.context)
                return true
            }
            getString(R.string.pref_key_blacklists) -> {
                SettingsActivity.startBlackListSettingsActivity(preference.context)
                return true
            }
            getString(R.string.pref_key_post_read) -> {
                SettingsActivity.startReadProgressSettingsActivity(preference.context)
                return true
            }
            getString(R.string.pref_key_backup) -> {
                SettingsActivity.startBackupSettingsActivity(preference.context)
                return true
            }
            getString(R.string.pref_key_network) -> {
                SettingsActivity.startNetworkSettingsActivity(preference.context)
                return true
            }
            getString(R.string.pref_key_check_update) -> {
                AppUpdate.checkUpdate(preference.context)
                return true
            }
            else -> return false
        }

    }
}
